core/drivers/crypto/se050/adaptors/utils/utils.c

```diff
@@ -0,0 +1,299 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) Foundries Ltd. 2020 - All Rights Reserved
+ * Author: Jorge Ramirez <jorge@foundries.io>
+ */
+
+#include <crypto/crypto.h>
+#include <se050.h>
+#include <string.h>
+
+#define BIT(nr) (1UL << (nr))
+
+/* base value for secure objects (transient and persistent) */
+#define OID_MIN			((uint32_t)(0x00000001))
+#define OID_MAX			((uint32_t)(OID_MIN + 0x7BFFFFFE))
+#define NBR_OID			((uint32_t)(OID_MAX - OID_MIN))
+
+#define IS_WATERMARKED(x)	(((x) & WATERMARKED(0)) == WATERMARKED(0))
+
+static void delete_transient_objects(void)
+{
+	Se05xSession_t *ctx = NULL;
+	uint8_t more = kSE05x_MoreIndicator_NA;
+	uint8_t *list = NULL;
+	size_t len = 1024;
+	smStatus_t status  = SM_NOT_OK;
+	uint16_t offset = 0;
+	uint32_t id = 0;
+	size_t nbr_pobj = 0, nbr_tobj = 0;
+	size_t i = 0;
+
+	if (!se050_session)
+		return;
+
+	ctx = &se050_session->s_ctx;
+
+	list = calloc(1, len);
+	if (!list)
+		return;
+	do {
+		status = Se05x_API_ReadIDList(ctx, offset, 0xFF, &more,
+					      list, &len);
+		if (status != SM_OK) {
+			free(list);
+			return;
+		}
+
+		offset = len;
+		for (i = 0; i < len; i += 4) {
+			id = (list[i + 0] << (3 * 8)) |
+			     (list[i + 1] << (2 * 8)) |
+			     (list[i + 2] << (1 * 8)) |
+			     (list[i + 3] << (0 * 8));
+
+			if (id >= OID_MAX || id == 0)
+				continue;
+
+			/* delete only transient objects */
+			if (id & BIT(0)) {
+				status = Se05x_API_DeleteSecureObject(ctx, id);
+				if (status != SM_OK) {
+					EMSG("Error erasing 0x%x", id);
+				} else {
+					nbr_tobj++;
+					DMSG("Erased 0x%x", id);
+				}
+			} else {
+				nbr_pobj++;
+			}
+		}
+	} while (more == kSE05x_MoreIndicator_MORE);
+
+	DMSG("Permanent objects in store %ld", nbr_pobj);
+	IMSG("Transient objects deleted  %ld", nbr_tobj);
+
+	free(list);
+}
+
+static uint32_t generate_oid(sss_key_object_mode_t mode)
+{
+	uint32_t oid = OID_MIN;
+	uint32_t random = 0;
+	size_t i = 0;
+
+	for (i = 0; i < NBR_OID; i++) {
+		if (crypto_rng_read(&random, sizeof(random)) != TEE_SUCCESS)
+			return 0;
+
+		random &= OID_MAX;
+
+		oid = OID_MIN + random;
+		if (oid > OID_MAX)
+			continue;
+
+		if (mode == kKeyObject_Mode_Transient)
+			oid |= BIT(0);
+		else
+			oid &= ~BIT(0);
+
+		if (!se050_key_exists(oid, &se050_session->s_ctx))
+			return oid;
+	}
+
+	return 0;
+}
+
+sss_status_t se050_get_oid(sss_key_object_mode_t mode, uint32_t *val)
+{
+	sss_status_t status = kStatus_SSS_Success;
+	uint16_t mem_t = 0, mem_p = 0;
+	uint32_t oid = 0;
+
+	if (!val)
+		return kStatus_SSS_Fail;
+
+	status = se050_get_free_memory(&se050_session->s_ctx, &mem_t,
+				       kSE05x_MemoryType_TRANSIENT_DESELECT);
+	if (status != kStatus_SSS_Success) {
+		mem_t = 0;
+		EMSG("failure retrieving transient free memory");
+		return kStatus_SSS_Fail;
+	}
+
+	status = se050_get_free_memory(&se050_session->s_ctx, &mem_p,
+				       kSE05x_MemoryType_PERSISTENT);
+	if (status != kStatus_SSS_Success) {
+		mem_p = 0;
```

Comment:
Same as `mem_t` this operation is not needed as we exits immediately.
In case function returns in error after, don't we need to free this memory and so `mem_t`?